### ISSUE TITLE
Explicit distinction between dummy and mock

### ISSUE THREAD

**Issue by zdenekdrahos** [phpspec/prophecy#223]
Hi,

I have run into problem with Prophecy's test doubles. My issue is not a bug, I guess it's a feature because the current behavior is described in [documentation](https://github.com/phpspec/prophecy#how-to-use-it). 

## Dummy vs mock in tests

When you develop some algorithm TDD-way then you end up with some dependencies and you are expecting their calls. Happy path uses all dependencies, but that's not always true for edge cases. Take a look at following simplified example:

```php
use Doctrine\ORM\EntityManagerInterface;

class Usecase
{
    private $entityManager;

    public function __construct(EntityManagerInterface $em)
    {
        $this->entityManager = $em;
    }

    public function execute($someCondition, Presenter $p)
    {
        if ($someCondition) {
            $this->entityManager->flush();
            $p->succeed();
        } else {
            $p->failed();
        }
    }
}

class UsecaseTest extends \PHPUnit_Framework_TestCase
{
    private $em;
    private $presenter;
    private $usecase;

    public function setUp()
    {
        $this->em = $this->prophesize(Doctrine\ORM\EntityManagerInterface::class);
        $this->presenter = $this->prophesize(Presenter::class);
        $this->usecase = new Usecase($this->em->reveal());
    }

    public function testFailure()
    {
        $this->presenter->failed()->shouldBeCalled();
        $this->usecase->execute(false, $this->presenter->reveal());
    }

    public function testSucceess()
    {
        $this->em->flush()->shouldBeCalled();
        $this->presenter->succeed()->shouldBeCalled();
        $this->usecase->execute(true, $this->presenter->reveal());
    }
}
```

This test suite has one big problem - **it can't be trusted. I can call `$entityManager` in `else branch` and test will pass.** I don't specify `$em` method prophecies in  `testFailure`, so `$em` is dummy which [won't throw UnexpectedCallException](https://github.com/phpspec/prophecy/blob/4f9b1eaf0a7da77c362f8d91cbc68ab1f4718d62/src/Prophecy/Call/CallCenter.php#L66). But I always need mock, honestly I rarely use dummies. Right now I would have to add one method `shouldNotBeCalled` in test so test would fail. But it still has disadvantages:

* test case is not readable - I don't care what is not called in specific edge case, but what is called
* tests can't be trusted - when I delete/add production code then test result should be different → I need feedback otherwise tests are useless

## Other mocking frameworks

I experimented with other tools ([source code](https://github.com/zdenekdrahos/php-mocks)). As you can see Prophecy is not the only one tool where not specified method is ignored:

| tool | test fails when unexpected method is called | fails when test is modified |
| ---- | --- | --- |
| prophecy | NO | `$d→unexpectedCall()→shouldNotBeCalled()` |
| phake | NO | `\Phake::verifyNoOtherInteractions($d)` |
| phpunit-mock | NO | never fails, [method is always ignored](https://github.com/sebastianbergmann/phpunit-mock-objects/issues/65) |
| mockery | **YES** | - |
| mockista | **YES** | - |

## Solution?

I miss explicit distinction between dummy and mock. I should be able specify if prophet is dummy/mock when prophet is created. In other words strict mode that forces `UnexpectedCallException` in dummy. Phake shows one solution. I think you could also extend current interface:

* `$this->prophesize(Presenter::class)->disableDummy()` 
* `$this->prophesize(Presenter::class, ProphetType::MOCK)`

**Comment by aik099** [comment]
> I miss explicit distinction between dummy and mock. I should be able specify if prophet is dummy/mock when prophet is created.

That's interesting. Not sure though if such behavior can be enabled system-wide, because it's a pain to pass some on/off switch to force mock creation on each created mock.

**Comment by zdenekdrahos** [comment]
Ok. I haven't studied prophecy internals. But I think that object with zero method prophecies can be

* **dummy** - every method returns `null`
* **mock** - doesn't expect any method call

AFAIK prophecy implements only dummy, but I need mock

**Comment by davedevelopment** [comment]
I'd actually call Prophecy's default a spy, I think dummies should throw if called, IMO dummies are used to satisfy typing and nothing else.

**Comment by stof** [comment]
@everzet I tend to agree with @davedevelopment here. Allowing any call with a ``null`` return (or a more complex return in case of PHP 7 with return type btw) looks weird to me, and I never got a use case where I wanted it (either no call should be done, or I should be aware of them). On the other hand, this behavior made me miss some calls in the tests while they would have been important.
what do you think about changing this behavior ? I will think about a way to make it gracefully
